staging/kos/pkg/controllers/subnet/validator.go

```diff
@@ -0,0 +1,442 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package subnet
+
+import (
+	"errors"
+	"fmt"
+	"net"
+	"sync"
+	"time"
+
+	"github.com/golang/glog"
+
+	k8scorev1api "k8s.io/api/core/v1"
+	k8serrors "k8s.io/apimachinery/pkg/api/errors"
+	k8smetav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	k8stypes "k8s.io/apimachinery/pkg/types"
+	k8sutilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	k8swait "k8s.io/apimachinery/pkg/util/wait"
+	k8scache "k8s.io/client-go/tools/cache"
+	k8sworkqueue "k8s.io/client-go/util/workqueue"
+
+	netv1a1 "k8s.io/examples/staging/kos/pkg/apis/network/v1alpha1"
+	kosclientv1a1 "k8s.io/examples/staging/kos/pkg/client/clientset/versioned/typed/network/v1alpha1"
+	netlistv1a1 "k8s.io/examples/staging/kos/pkg/client/listers/network/v1alpha1"
+	kosctlrutils "k8s.io/examples/staging/kos/pkg/controllers/utils"
+)
+
+// TODO: something which makes the validation failure visible to the client
+// (event, Prometheus, status field, all of them, etc...).
+
+// usable is the value the validator sets a subnet's status.ValidationOutcome to
+// if there are no conflicts with other subnets.
+const usable = "usable"
+
+// conflictsCache holds information for one subnet regarding conflicts with
+// other subnets. There's no guarantee that the cache is up-to-date: a subnet Y
+// stored in it might no longer be in conflict with the owning subnet, for
+// instance because of an update to Y's CIDR following its addition to the
+// cache.
+type conflictsCache struct {
+	// ownerData stores the data relevant to validation for the subnet owning
+	// the conflicts cache.
+	ownerData *subnetData
+
+	// rivals is the list of namespaced names of the subnets that were observed
+	// to conflict with the subnet owning the conflicts cache. If subnet X is
```

Comment:
This description is backwards, X and Y are reversed.

Reply:
c2111fb replaces the original comment with your suggestion.